### PR TITLE
[SPARK-18579][SQL] Use ignoreLeadingWhiteSpace and ignoreTrailingWhiteSpace options in CSV writing

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to support _not_ trimming the white spaces when writing out. These are `false` by default in CSV reading path but these are `true` by default in CSV writing in univocity parser.

Both `ignoreLeadingWhiteSpace` and `ignoreTrailingWhiteSpace` options are not being used for writing and therefore, we are always trimming the white spaces.

It seems we should provide a way to keep this white spaces easily.

WIth the data below:

```scala
val df = spark.read.csv(Seq("a , b  , c").toDS)
df.show()
```

```
+---+----+---+
|_c0| _c1|_c2|
+---+----+---+
| a | b  |  c|
+---+----+---+
```

**Before**

```scala
df.write.csv("/tmp/text.csv")
spark.read.text("/tmp/text.csv").show()
```

```
+-----+
|value|
+-----+
|a,b,c|
+-----+
```

It seems this can't be worked around via `quoteAll` too.

```scala
df.write.option("quoteAll", true).csv("/tmp/text.csv")
spark.read.text("/tmp/text.csv").show()
```
```
+-----------+
|      value|
+-----------+
|"a","b","c"|
+-----------+
```

**After**

```scala
df.write.option("ignoreLeadingWhiteSpace", false).option("ignoreTrailingWhiteSpace", false).csv("/tmp/text.csv")
spark.read.text("/tmp/text.csv").show()
```

```
+----------+
|     value|
+----------+
|a , b  , c|
+----------+
```

Note that this case is possible in R

```r
> system("cat text.csv")
f1,f2,f3
a , b  , c
> df <- read.csv(file="text.csv")
> df
  f1   f2 f3
1 a   b    c
> write.csv(df, file="text1.csv", quote=F, row.names=F)
> system("cat text1.csv")
f1,f2,f3
a , b  , c
```

## How was this patch tested?

Unit tests in `CSVSuite` and manual tests for Python.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala [hunk 1]
```diff
@@ -92,8 +92,13 @@ class CSVOptions(
...
   val inferSchemaFlag = getBool("inferSchema")
-  val ignoreLeadingWhiteSpaceFlag = getBool("ignoreLeadingWhiteSpace")
-  val ignoreTrailingWhiteSpaceFlag = getBool("ignoreTrailingWhiteSpace")
+  val ignoreLeadingWhiteSpaceInRead = getBool("ignoreLeadingWhiteSpace", default = false)
+  val ignoreTrailingWhiteSpaceInRead = getBool("ignoreTrailingWhiteSpace", default = false)
+
+  // For write, both options were `true` by default. We leave it as `true` for
+  // backwards compatibility.
+  val ignoreLeadingWhiteSpaceFlagInWrite = getBool("ignoreLeadingWhiteSpace", default = true)
+  val ignoreTrailingWhiteSpaceFlagInWrite = getBool("ignoreTrailingWhiteSpace", default = true)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: This case has been `true` by defaults before (https://github.com/uniVocity/univocity-parsers/blob/644e76005f52a3a2b9b29844c3ed9daa5ba128aa/src/main/java/com/univocity/parsers/common/CommonSettings.java#L182).

So, I left them as `true` by default for now.

**Comment by felixcheung** [comment]
is this a behavior change from 2.0/2.1?

**Comment by HyukjinKwon** [comment]
Oh, no. I left them with `true` by default to keep the original behaviour. It could look a bit odd because it is `false` in read and `true` in write by default but I tried my best to keep the original behaviour by this.

File: python/pyspark/sql/readwriter.py [hunk 2]
```diff
@@ -748,13 +748,21 @@ def csv(self, path, mode=None, compression=None, sep=None, quote=None, escape=No
                                 formats follow the formats at ``java.text.SimpleDateFormat``.
                                 This applies to timestamp type. If None is set, it uses the
                                 default value, ``yyyy-MM-dd'T'HH:mm:ss.SSSZZ``.
+        :param ignoreLeadingWhiteSpace: defines whether or not leading whitespaces from values
```
Comment (felixcheung) [inline python/pyspark/sql/readwriter.py]: perhaps to start like `:param escapeQuotes: A flag indicating whether `?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 3]
```diff
@@ -581,6 +581,10 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
    * <li>`timestampFormat` (default `yyyy-MM-dd'T'HH:mm:ss.SSSZZ`): sets the string that
    * indicates a timestamp format. Custom date formats follow the formats at
    * `java.text.SimpleDateFormat`. This applies to timestamp type.</li>
+   * <li>`ignoreLeadingWhiteSpace` (default `true`): defines whether or not leading whitespaces
```
Comment (felixcheung) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: ditto, `* <li>`quoteAll` (default `false`): A flag indicating whether `

File: python/pyspark/sql/readwriter.py [hunk 4]
```diff
@@ -748,13 +748,21 @@ def csv(self, path, mode=None, compression=None, sep=None, quote=None, escape=No
...
+                                        being written should be skipped. If None is set, it uses
+                                        the default value, ``true``.
+        :param ignoreTrailingWhiteSpace: defines whether or not trailing whitespaces from values
+                                         being written should be skipped. If None is set, it uses
+                                         the default value, ``true``.
```
Comment (felixcheung) [inline python/pyspark/sql/readwriter.py]: any existing python test to add testing these new options?

**Review by felixcheung — APPROVED** [review]
LGTM

**Comment by felixcheung** [comment]
merged to master.

**Comment by HyukjinKwon** [comment]
Thank you @felixcheng.

**Comment by prabcs** [comment]
👋 Can this change be backported to 2.1 ?

**Comment by felixcheung** [comment]
This is an API change which we generally don't backport to a patch release. Can you elaborate on why you need it?

**Comment by prabcs** [comment]
We're using 2.1.1 as of now and 2.2 seems a bit far; Hence, exploring the options of backporting this to 2.1 somehow.
Also, directly patching this on 2.1 at our end is non-trivial as there are some more code changes between 2.1 and this version.

**Comment by felixcheung** [comment]
actually, 2.2 is likely getting released in weeks, while there might not be a 2.1.2 release..

**Comment by prabcs** [comment]
OK, great then !! We'll use 2.2
Thanks !

**Pull request closed without merging**